core/src/test/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitorSpec.scala

```diff
@@ -443,6 +493,102 @@ class AvroUpsertMonitorSpec(_system: ActorSystem) extends ApiServiceSpec with Mo
     }
   }
 
+  it should "ack pubsub message if upsert published to nonexistent workspace" in withTestDataApiServices { services =>
+    val timeout = 30000 milliseconds
+    val interval = 250 milliseconds
+    val importId1 = UUID.randomUUID()
+
+    // add the imports and their statuses to the mock importserviceDAO
+    val mockImportServiceDAO =  setUp(services)
+    mockImportServiceDAO.imports += (importId1 -> ImportStatuses.ReadyForUpsert)
+
+    val contents = makeOpsJsonString(100)
+
+    // Store upsert json file
+    Await.result(googleStorage.createBlob(bucketName, GcsBlobName(importId1.toString), contents.getBytes()).compile.drain.unsafeToFuture(), Duration.apply(10, TimeUnit.SECONDS))
+
+    // Make sure the file saved properly
+    Await.result(googleStorage.unsafeGetBlobBody(bucketName, GcsBlobName(importId1.toString)).unsafeToFuture(), Duration.apply(10, TimeUnit.SECONDS))
+
+    // Publish message on the request topic with a nonexistent workspace
+    val messageAttributes = testAttributes(importId1) ++ Map("workspaceName" -> "does_not_exist")
+    services.gpsDAO.publishMessages(importReadPubSubTopic, List(MessageRequest(importId1.toString, messageAttributes)))
+
+    // check if correct message was posted on request topic. This will start the upsert attempt.
+    eventually(Timeout(scaled(timeout)), Interval(scaled(interval))) {
+      assert(services.gpsDAO.receivedMessage(importReadPubSubTopic, importId1.toString))
+    }
+
+    // upsert will fail; check that a pubsub message was acked.
+    eventually(Timeout(scaled(timeout)), Interval(scaled(interval))) {
+      assert(!services.gpsDAO.acks.isEmpty)
+    }
```

Comment:
1) Since there's no good way to verify that we've acked the correct message - we don't have access to the ackId - how about at least adding an assertion earlier in the test to verify that `acks` is empty? Something like:
```scala
    // acks should be empty at this point
    eventually(Timeout(scaled(timeout)), Interval(scaled(interval))) {
      services.gpsDAO.acks shouldBe empty
    }
```
before publishing the "start import" message to the request topic. (this applies to all the new tests in this PR)

2) To be Scala/Scalatest idiomatic, check out [matchers](https://www.scalatest.org/user_guide/using_matchers#checkingForEmptiness). You could change `assert(!services.gpsDAO.acks.isEmpty)` to `services.gpsDAO.acks should not be empty`. This is functionally equivalent and doesn't make a difference to test behavior itself, but might be easier to read and more consistent with other tests.

Reply:
whoops, this second assertion, after the upsert job runs, needs to be:
```scala
services.gpsDAO.acks should not be empty
```

Reply:
It should be empty before publishmessage and after receive message.  But are you saying I should check that it's *not* empty inbetween?

Reply:
Ok, did it and double-checked that it fails without the code change.